src/bvh/ordered_traverse.rs

```diff
@@ -0,0 +1,345 @@
+        match self.bvh.nodes[self.node_index] {
+            BvhNode::Node {
+                child_r_index,
+                child_l_index,
+                ..
+            } => match rest_child {
+                RestChild::Left => {
+                    self.node_index = child_l_index;
+                    self.has_node = true;
+                }
+                RestChild::Right => {
+                    self.node_index = child_r_index;
+                    self.has_node = true;
+                }
+                RestChild::None => {
+                    self.has_node = false;
+                }
+            },
+            BvhNode::Leaf { .. } => {
+                self.has_node = false;
+            }
+        }
+    }
+}
+
+impl<'bvh, 'shape, T, const D: usize, Shape: Bounded<T, D>, const ASCENDING: bool> Iterator
+    for DistanceTraverseIterator<'bvh, 'shape, T, D, Shape, ASCENDING>
+where
+    T: Scalar + Copy + RealField,
+{
+    type Item = &'shape Shape;
+
+    fn next(&mut self) -> Option<&'shape Shape> {
+        loop {
+            if self.is_stack_empty() && !self.has_node {
+                // Completed traversal.
+                break;
+            }
+
+            if self.has_node {
+                // If we have any node, attempt to move to its nearest child.
+                let stack_info = if ASCENDING {
+                    self.move_nearest()
+                } else {
+                    self.move_furthest()
+                };
+                // Save current node and farthest child
+                self.stack_push(stack_info)
+            } else {
+                // Go back up the stack and see if a node or leaf was pushed.
+                let (node_index, rest_child) = self.stack_pop();
+                self.node_index = node_index;
+                match self.bvh.nodes[self.node_index] {
+                    BvhNode::Node { .. } => {
+                        // If a node was pushed, now move to `unvisited` rest child, next in order.
+                        self.move_rest(rest_child);
+                    }
+                    BvhNode::Leaf { shape_index, .. } => {
+                        // We previously pushed a leaf node. This is the "visit" of the in-order traverse.
+                        // Next time we call `next()` we try to pop the stack again.
+                        return Some(&self.shapes[shape_index]);
+                    }
+                }
+            }
+        }
+        None
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::aabb::Bounded;
+    use crate::bvh::Bvh;
+    use crate::ray::Ray;
+    use crate::testbase::{generate_aligned_boxes, TBvh3, TPoint3, TVector3, UnitBox};
+    use std::collections::HashSet;
+
+    /// Create a `Bvh` for a fixed scene structure.
+    pub fn build_some_bvh() -> (Vec<UnitBox>, TBvh3) {
+        let mut boxes = generate_aligned_boxes();
+        let bvh = Bvh::build(&mut boxes);
+        (boxes, bvh)
+    }
+
+    fn traverse_and_verify_oredered(
```

Comment:
```suggestion
    fn traverse_distance_and_verify_order(
```

Typo, and also a rename here would be helpful